docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,521 @@
+- The Machine must have conditions BootstrapReady.
+- The Kubernetes CSR spec has the needed groups
+- The Kubernetes CSR spec is limited to needed usages (e.g. client auth)
+- The Kubernetes CSR spec is limited to needed extensions (e.g. no CA extension)
+- Parse the CSR and verify that the CN is the same as .spec.username
+- Parse the CSR and verify that the Organization is the same as .spec.Groups
+- Parse the CSR and ensure that no SANs are appended for kubelet client certificates
+
+#### CSR format used by kubelet-authenticator
+We propose the introduction of X.509 extension attributes based
+on those reserved for the Kubernetes GCP cloud provider within Google’s organization ID allocation.
+
+We will request via SIG Architecture or CNCF to apply for an [IANA OID registration
+block][iana-issue] for the Kubernetes project.
+
+##### OIDs
+
+* **OID Suffix**: 2.1.21
+* **Name**: KubernetesNodeProviderIdentifierOID
+* **Description**: An identifier for the machine, should be the same or a derivative of the node
+  provider ID. This is the equivalent of Google’s CloudComputeInstanceIdentifierOID, which we can
+  reuse for a proof of concept (1.3.6.1.4.1.11129.2.1.21).
+
+#### CSR PEM Blocks
+
+The following blocks will be added to CSRs following Section 2 of [RFC7468].
+
+| Block Name                                 | Description                                          |
+| ------------------------------------------ | ---------------------------------------------------- |
+| KUBELET AUTHENTICATOR ATTESTATION PROVIDER | string describing the attestation provider           |
+| KUBELET AUTHENTICATOR ATTESTATION DATA     | the actual attestation data to perform validation on |
+
+#### Attestation data
+
+Attestation data will be appended with the following headers and footers and MUST
+be base64 encoded.
+
+Example CSR:
+```
+-----BEGIN ATTESTATION DATA-----
+S25vd2luZyBtZSBBbGFuIFBhcnRyaWRnZSwga25vd2luZyB5b3UgS3ViZXJuZXRlcyBjbHVzdGVyLCBhaGEh
+-----END ATTESTATION DATA-----
+```
+The format of the attestation block is left to the provider.
+
+#### Core Specification
+- Core Cluster API MUST provide the following implementations of CSRs and signers:
+  - `cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure` which implement an “Always Allow” type
+    signer that provides equivalent security to Cluster API v1alpha3. This is only to be used for
+    providers where no secure mechanism exists.
+
+- Core Cluster API MIGHT provide the following implementations of CSRs and signers:
+  - `cluster.x-k8s.io/kube-apiserver-client-kubelet-tpm` and `cluster-x-k8s-io/kubelet-serving-tpm`
+    - Will implement TPM-based certificate signers and requesters based on the cloud-provider-gcp
+    implementation. - However, since the mechanism for retrieving endorsement keys varies across
+    platforms, the TPM signer will additionally require a provider specific mechanism to provide the
+    TPM Endorsement Key.
```

Comment:
IIUC, the endorsement key can't be used for signing. You need to use the attestation key. To prove that the node has the expected EK, you need a challenge-response protocol.

Reply:
Taking a look at the Bloomberg SPIRE plugin. The challenge response model seems a good improvement on the K8s GCP Cloud Provider and we have a way to do it I think.

Reply:
The text should have read `the TPM signer will additionally require a provider specific mechanism to provide the TPM Endorsement Key's Certificate Authority`, but yes, following the GCP implementation, we'd be generating an AIK (links to both the GCP and SPIRE implementations have been added). We're not going to implement TPM right away, the POC will start with the AWS and Azure attestors to fix ASG implementations, so a more detailed TPM proposal will follow.